from kinetic_datanator.data_source import array_express
import unittest
import tempfile
import natsort
import os
import shutil
from kinetic_datanator.data_source import download_ax
import collections

class QuickTestArrayExpress(unittest.TestCase):

	@classmethod
	def setUpClass(cls):
		cls.cache_dirname = tempfile.mkdtemp()
		src = array_express.ArrayExpress(cache_dirname=cls.cache_dirname, download_backup=False, load_content=False)
		src.session.close()
		src.engine.dispose()
		download_ax.download_all_metadata(start_year=2001, end_year=2001)
		
	@classmethod
	def tearDownClass(cls):
		shutil.rmtree(cls.cache_dirname)
		shutil.rmtree("./AllExperiments")
		shutil.rmtree("./AllSamples")

	def setUp(self):
		self.src = array_express.ArrayExpress(cache_dirname=self.cache_dirname, download_backup=False, load_content=False)

	def test_load_all_metadata(self):
		src = self.src
		session = src.session

		src.load_experiments_from_text()
		src.load_samples_from_text()
		q = session.query(array_express.Experiment)
		self.assertEqual(q.count(), 3)
		self.checkEqual([c.name for c in q.all()], [u'eye-SAGE', u'Topoisomerase function in E. coli replication fork movement', u'Genetic diversity among Helicobacter pylori strains'])
		self.checkEqual([c.name for c in q.all()], [u'eye-SAGE', u'Topoisomerase function in E. coli replication fork movement', u'Genetic diversity among Helicobacter pylori strains'])
		self.checkEqual([c.description for c in q.all()], [u'Human retinal and RPE SAGE libraries. Profile of the genes expressed in the human peripheral retina, macula, and retinal pigment epithelium determined through serial analysis of gene expression (SAGE). Keywords: other', u'Analysis of topoisomerase function in bacterial replication fork movement: use of DNA microarrays. We used DNA microarrays of the Escherichia coli genome to trace the progression of chromosomal replication forks in synchronized cells. We found that both DNA gyrase and topoisomerase IV (topo IV) promote replication fork progression. When both enzymes were inhibited, the replication fork stopped rapidly. The elongation rate with topo IV alone was 1/3 of normal. Genetic data confirmed and extended these results. Inactivation of gyrase alone caused a slow stop of replication. Topo IV activity was sufficient to prevent accumulation of (+) supercoils in plasmid DNA in vivo, suggesting that topo IV can promote replication by removing (+) supercoils in front of the chromosomal fork. This study is detailed in Khodursky AB et al.(2000) Proc Natl Acad Sci U S A 97:9419-24 Keywords: other', u"Helicobacter pylori colonizes the stomach of half of the world's population, causing a wide spectrum of disease ranging from asymptomatic gastritis to ulcers to gastric cancer. Although the basis for these diverse clinical outcomes is not understood, more severe disease is associated with strains harboring a pathogenicity island. To characterize the genetic diversity of more and less virulent strains, we examined the genomic content of 15 H. pylori clinical isolates by using a whole genome H. pylori DNA microarray. We found that a full 22% of H. pylori genes are dispensable in one or more strains, thus defining a minimal functional core of 1281 H. pylori genes. While the core genes encode most metabolic and cellular processes, the strain-specific genes include genes unique to H. pylori, restriction modification genes, transposases, and genes encoding cell surface proteins, which may aid the bacteria under specific circumstances during their long-term infection of genetically diverse hosts. We observed distinct patterns of the strain-specific gene distribution along the chromosome, which may result from different mechanisms of gene acquisition and loss. Among the strain-specific genes, we have found a class of candidate virulence genes identified by their coinheritance with the pathogenicity island."])
		self.checkEqual([ob.name for thing in [c.organism_ for c in q.all()] for ob in thing], [u'Homo sapiens', u'Escherichia coli', u'Helicobacter pylori'])
		self.checkEqual([c.submission_date for c in q.all()], [u'2001-10-02', u'2001-07-29', u'2001-07-24'])
		self.checkEqual([c.release_date for c in q.all()], [u'2001-10-03', u'2001-07-30', u'2001-07-25'])

		q = session.query(array_express.ExperimentDesign)

		q = session.query(array_express.ExperimentType)
		self.checkEqual([c.name for c in q.all()], [u'transcription profiling by SAGE', u'other', u'unknown experiment type'])

		q = session.query(array_express.Sample)
		self.checkEqual([c.name for c in q.all()], [u'GSM505 2', u'GSM505 1', u'GSM504 1', u'GSM504 2', u'GSM503 2', u'GSM503 1', u'GSM502 1', u'GSM502 2', u'GSM501 1', u'GSM501 2', u'GSM500 1', u'GSM500 2', u'GSM499 1', u'GSM499 2', u'GSM498 1', u'GSM498 2', u'GSM497 2', u'GSM497 1', u'GSM496 2', u'GSM496 1', u'GSM495 1', u'GSM495 2', u'GSM494 1', u'GSM494 2', u'GSM493 2', u'GSM493 1', u'GSM492 2', u'GSM492 1', u'GSM491 2', u'GSM491 1', u'GSM490 2', u'GSM490 1', u'GSM489 2', u'GSM489 1', u'GSM488 1', u'GSM488 2', u'GSM487 1', u'GSM487 2', u'GSM486 2', u'GSM486 1', u'GSM485 1', u'GSM485 2', u'GSM484 1', u'GSM484 2', u'GSM483 2', u'GSM483 1', u'GSM482 2', u'GSM482 1', u'GSM481 2', u'GSM481 1', u'GSM480 2', u'GSM480 1', u'GSM479 2', u'GSM479 1', u'GSM478 2', u'GSM478 1', u'GSM477 1', u'GSM477 2', u'GSM476 1', u'GSM476 2', u'GSM475 2', u'GSM475 1', u'GSM474 2', u'GSM474 1', u'GSM473 1', u'GSM473 2', u'GSM472 2', u'GSM472 1', u'GSM471 1', u'GSM471 2', u'GSM470 1', u'GSM470 2', u'GSM574 1', u'GSM573 1', u'GSM572 1', u'GSM571 1', u'GSM569 2', u'GSM569 1', u'GSM568 2', u'GSM568 1', u'GSM567 2', u'GSM567 1', u'GSM566 1', u'GSM566 2', u'GSM554 2', u'GSM554 1', u'GSM553 2', u'GSM553 1', u'GSM552 2', u'GSM552 1', u'GSM551 1', u'GSM551 2', u'GSM550 1', u'GSM550 2', u'GSM530 2', u'GSM530 1', u'GSM529 2', u'GSM529 1', u'GSM528 2', u'GSM528 1', u'GSM527 1', u'GSM527 2', u'GSM526 1', u'GSM526 2', u'GSM525 1', u'GSM525 2', u'GSM524 2', u'GSM524 1', u'GSM523 1', u'GSM523 2', u'GSM522 2', u'GSM522 1', u'GSM521 1', u'GSM521 2', u'GSM520 2', u'GSM520 1', u'GSM518 1', u'GSM518 2'])
		self.checkEqual([c.assay for c in q.all()], [u'GSM505', u'GSM505', u'GSM504', u'GSM504', u'GSM503', u'GSM503', u'GSM502', u'GSM502', u'GSM501', u'GSM501', u'GSM500', u'GSM500', u'GSM499', u'GSM499', u'GSM498', u'GSM498', u'GSM497', u'GSM497', u'GSM496', u'GSM496', u'GSM495', u'GSM495', u'GSM494', u'GSM494', u'GSM493', u'GSM493', u'GSM492', u'GSM492', u'GSM491', u'GSM491', u'GSM490', u'GSM490', u'GSM489', u'GSM489', u'GSM488', u'GSM488', u'GSM487', u'GSM487', u'GSM486', u'GSM486', u'GSM485', u'GSM485', u'GSM484', u'GSM484', u'GSM483', u'GSM483', u'GSM482', u'GSM482', u'GSM481', u'GSM481', u'GSM480', u'GSM480', u'GSM479', u'GSM479', u'GSM478', u'GSM478', u'GSM477', u'GSM477', u'GSM476', u'GSM476', u'GSM475', u'GSM475', u'GSM474', u'GSM474', u'GSM473', u'GSM473', u'GSM472', u'GSM472', u'GSM471', u'GSM471', u'GSM470', u'GSM470', u'GSM574', u'GSM573', u'GSM572', u'GSM571', u'GSM569', u'GSM569', u'GSM568', u'GSM568', u'GSM567', u'GSM567', u'GSM566', u'GSM566', u'GSM554', u'GSM554', u'GSM553', u'GSM553', u'GSM552', u'GSM552', u'GSM551', u'GSM551', u'GSM550', u'GSM550', u'GSM530', u'GSM530', u'GSM529', u'GSM529', u'GSM528', u'GSM528', u'GSM527', u'GSM527', u'GSM526', u'GSM526', u'GSM525', u'GSM525', u'GSM524', u'GSM524', u'GSM523', u'GSM523', u'GSM522', u'GSM522', u'GSM521', u'GSM521', u'GSM520', u'GSM520', u'GSM518', u'GSM518'])
		self.checkEqual([ob.name for thing in [c.characteristics for c in q.all()] for ob in thing],[u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism'])

		q = session.query(array_express.Extract)
		self.checkEqual([c.sample_id for c in q.all()], [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118])
		self.checkEqual([c.name for c in q.all()], [u'GSM505 extract 2', u'GSM505 extract 1', u'GSM504 extract 1', u'GSM504 extract 2', u'GSM503 extract 2', u'GSM503 extract 1', u'GSM502 extract 1', u'GSM502 extract 2', u'GSM501 extract 1', u'GSM501 extract 2', u'GSM500 extract 1', u'GSM500 extract 2', u'GSM499 extract 1', u'GSM499 extract 2', u'GSM498 extract 1', u'GSM498 extract 2', u'GSM497 extract 2', u'GSM497 extract 1', u'GSM496 extract 2', u'GSM496 extract 1', u'GSM495 extract 1', u'GSM495 extract 2', u'GSM494 extract 1', u'GSM494 extract 2', u'GSM493 extract 2', u'GSM493 extract 1', u'GSM492 extract 2', u'GSM492 extract 1', u'GSM491 extract 2', u'GSM491 extract 1', u'GSM490 extract 2', u'GSM490 extract 1', u'GSM489 extract 2', u'GSM489 extract 1', u'GSM488 extract 1', u'GSM488 extract 2', u'GSM487 extract 1', u'GSM487 extract 2', u'GSM486 extract 2', u'GSM486 extract 1', u'GSM485 extract 1', u'GSM485 extract 2', u'GSM484 extract 1', u'GSM484 extract 2', u'GSM483 extract 2', u'GSM483 extract 1', u'GSM482 extract 2', u'GSM482 extract 1', u'GSM481 extract 2', u'GSM481 extract 1', u'GSM480 extract 2', u'GSM480 extract 1', u'GSM479 extract 2', u'GSM479 extract 1', u'GSM478 extract 2', u'GSM478 extract 1', u'GSM477 extract 1', u'GSM477 extract 2', u'GSM476 extract 1', u'GSM476 extract 2', u'GSM475 extract 2', u'GSM475 extract 1', u'GSM474 extract 2', u'GSM474 extract 1', u'GSM473 extract 1', u'GSM473 extract 2', u'GSM472 extract 2', u'GSM472 extract 1', u'GSM471 extract 1', u'GSM471 extract 2', u'GSM470 extract 1', u'GSM470 extract 2', u'GSM574 extract 1', u'GSM573 extract 1', u'GSM572 extract 1', u'GSM571 extract 1', u'GSM569 extract 2', u'GSM569 extract 1', u'GSM568 extract 2', u'GSM568 extract 1', u'GSM567 extract 2', u'GSM567 extract 1', u'GSM566 extract 1', u'GSM566 extract 2', u'GSM554 extract 2', u'GSM554 extract 1', u'GSM553 extract 2', u'GSM553 extract 1', u'GSM552 extract 2', u'GSM552 extract 1', u'GSM551 extract 1', u'GSM551 extract 2', u'GSM550 extract 1', u'GSM550 extract 2', u'GSM530 extract 2', u'GSM530 extract 1', u'GSM529 extract 2', u'GSM529 extract 1', u'GSM528 extract 2', u'GSM528 extract 1', u'GSM527 extract 1', u'GSM527 extract 2', u'GSM526 extract 1', u'GSM526 extract 2', u'GSM525 extract 1', u'GSM525 extract 2', u'GSM524 extract 2', u'GSM524 extract 1', u'GSM523 extract 1', u'GSM523 extract 2', u'GSM522 extract 2', u'GSM522 extract 1', u'GSM521 extract 1', u'GSM521 extract 2', u'GSM520 extract 2', u'GSM520 extract 1', u'GSM518 extract 1', u'GSM518 extract 2'])

		q = session.query(array_express.Variable)
		#print [c.name for c in q.all()]
		#print [c.value for c in q.all()]
		#print [c.unit for c in q.all()]

		q = session.query(array_express.Characteristic)
		self.checkEqual([c.name for c in q.all()], [u'Organism', u'Organism', u'Organism'])
		self.checkEqual([c.value for c in q.all()], [u'Helicobacter pylori', u'Homo sapiens', u'Escherichia coli'])

	def checkEqual(self, L1, L2):
		self.assertEqual(sorted(L1), sorted(L2))


class LongTestArrayExpress(unittest.TestCase):

	@classmethod
	def setUpClass(cls):
		cls.cache_dirname = tempfile.mkdtemp()
		src = array_express.ArrayExpress(cache_dirname=cls.cache_dirname, download_backup=False, load_content=False)
		src.session.close()
		src.engine.dispose()
		download_ax.download_all_metadata(start_year=2001, end_year=2002)
		
	@classmethod
	def tearDownClass(cls):
		shutil.rmtree(cls.cache_dirname)
		shutil.rmtree("./AllExperiments")
		shutil.rmtree("./AllSamples")

	def setUp(self):
		self.src = array_express.ArrayExpress(cache_dirname=self.cache_dirname, download_backup=False, load_content=False)


	def test_load_all_metadata(self):
		src = self.src
		session = src.session

		src.load_experiments_from_text()
		src.load_samples_from_text()
		q = session.query(array_express.Experiment)
		self.assertEqual( q.count(), 26)
		#self.checkEqual([c.name for c in q.all()], [u'Proteomic view of the Plasmodium falciparum life cycle', u'Bovine Glomerular and Aortic Endothelial Cells', u'Embryo development of Medicago truncatula', u'Transcription profiling of fission yeast meiosis and sporulation', u'Transcription profiling of fission yeast meiosis and sporulation', u'Transcription profiling of fission yeast meiosis and sporulation', u'Transcription profiling of fission yeast meiosis and sporulation', u'Transcription profiling of fission yeast meiosis and sporulation', u'Transcription profiling of fission yeast sexual differentiation', u'P19 embryonic carcinoma (EC) cells induced to form cardiomyocytes in vitro', u'mRNAs translated at eIF4F', u'Molecular portraits of human breast tumors', u'Diffuse large B-cell lymphoma', u'Human mammary epithelium and breast cancer', u'Antibody Array Figures', u'DNA copy-number changes', u'Secreted and membrane-associated genes', u'Stresses induce gene expression in Arabidopsis', u'Transcription profiling of human cell lines and tissues (Novartis/GNF)', u'SAGE profiles from cultured keratinocytes and human epidermis', u'Adaptive evolution in yeast', u'cdc15 block-release', u'Manipulation in phospate levels', u'eye-SAGE', u'Topoisomerase function in E. coli replication fork movement', u'Genetic diversity among Helicobacter pylori strains'])
		self.checkEqual([c.description for c in q.all()], [u'The completion of the Plasmodium falciparum clone 3D7 genome provides a basis on which to conduct comparative proteomics studies of this human pathogen. Here, we applied a high-throughput proteomics approach to identify new potential drug and vaccine targets and to better understand the biology of this complex protozoan parasite. We characterized four stages of the parasite life cycle (sporozoites, merozoites, trophozoites and gametocytes) by multidimensional protein identification technology. Functional profiling of over 2,400 proteins agreed with the physiology of each stage. Unexpectedly, the antigenically variant proteins of var and rif genes, defined as molecules on the surface of infected erythrocytes, were also largely expressed in sporozoites. The detection of chromosomal clusters encoding co-expressed proteins suggested a potential mechanism for controlling gene expression. Keywords: ordered', u'SAGE libraries were prepared from mRNA of cultured bovine glomerular and aortic endothelial cells. Cells both cell lines were positive for acetylated LDL uptake and vWF expression, and were proven to be free of non-endothelial cells.  Cultures were studied at passage 9-10. The cells were confluent, cultured in parallel, under identical conditions. Cells were plated on gelatin coated cell culture plastic in RPMI 1640 medium containing 10% FBS and streptomycin/penicillin. The last medium change was 48 hours prior to mRNA harvest. Keywords = endothelial, glomerular, kidney, aorta, Keywords: other', u'Early stage of embryo development from leaf-derived embryogenic callus of M. truncatula.', u'Transcriptional profile of fission yeast meiosis and sporulation. Fission yeast cells undergo meiosis and sporulation under conditions of nutritional stress, most frequently nitrogen starvation.  This is a complex developmental process, which results in the formation of four spores that are highly resistant to environmental stress. We have carried out time courses of diploid fission yeast cells undergoing meiosis and sporulation. To achieve the best possible synchrony we have used thermosensitive mutants in the meiotic inhibitor pat1. We first pre-synchronized the cells in G1 by removal of nitrogen at the permissive temperature for pat1, and then induced meiosis by inactivating pat1 with a temperature shift. pat1-induced meiosis is a standard technique, resulting in a highly synchronous meiosis that is very similar to the normal process. The use of pat1 mutants and a temperature-shift may cause some artifacts, so these results should be viewed together with temperature shift controls and especially with the results of the time course done with wild type diploids, available under accession numbers E-SNGR-3 to E-SNGR-7. In all cases we used as reference a pool consisting of equal amounts of RNA  from all time points of the experiment. This experiment describes the first biological replicate for pat1 using array design A-SNGR-2. Note that there are three other biological replicates, which use array design A-SNGR-1 in experiment E-SNGR-2.', u'pat1 meiosis is very synchronous (see E-SNGR-2), but the use of a temperature shift and the prolonged nitrogen starvation can lead to artifacts. As a control for the pat1 experiment,non-sporulating pat1+ diploids were treated exactly as in the protocol for pat1 cells, with the exception that no nitrogen was added at the time of the temperature shift. Under these conditions the cells remain blocked in G1.  As a reference we used a pool consisting of equal amount of RNA from all time points of the experiments. This experiment is part of a series, E-SNGR-2 to E-SNGR-7.', u'pat1 meiosis is very synchronous (see E-SNGR-2), but the use of a temperature shift and the prolonged nitrogen starvation can lead to artifacts. As a control for the E-SNGR-2 experiment, non-sporulating pat1+ diploids were treated exactly as in the protocol for E-SNGR-2  cells. Under these conditions the cells re-enter the mitotic cycle.  As a reference we used a pool consisting of equal amount of RNA from all time points of the experiments. This experiment is part of a series, E-SNGR-2 to E-SNGR-7.', u'Study of the meiotic transcriptional profile of transcription factor mutants. The transcription factors atf21 and atf31 are highly induced during the meiotic divisions and required for proper spore formation. To identify possible targets of these transcription factors we compared the transcriptional profile of atf21 and atf31 mutants with that of wild type cells undergoing meiosis.  We used haploid h90 cells for this experiment. Removal of nitrogen from the medium induces mating to form diploids, which immediately undergo meiosis and sporulation. This type of meiosis is not very synchronous. This experiment is part of a series, E-SNGR-2 to E-SNGR-7.', u'Fission yeast cells undergo meiosis and sporulation under conditions of nutritional stress, most frequently nitrogen starvation. This is a complex developmental process, which results in the formation of four spores that are highly resistant to environmental stress. We have carried out time course experiments with diploid fission yeast cells undergoing meiosis and sporulation. In this experiment we have used wild type diploids, and induced meiosis by removal of nitrogen from the medium. As a reference we used a pool consisting of equal amount of RNA from all time points of the experiments. This experiment is part of a series, with accession numbers E-SNGR-2 to E-SNGR-7.', None, u'SAGE identification of differentiation responsive genes in P19 embryonic cells induced to form cardiomyocytes in vitro. P19 embryonic carcinoma (EC) cells, induced to form cardiomyocytes in vitro - undifferentiated cells, day 3+0.5 and day 3+3.0 of differentiation protocol. Keywords = EC cells, P19, differentiation, cardiomyocytes Keywords: time-course', u"Identification of eukaryotic mRNAs that are translated at reduced cap binding complex eIF4F concentrations using a cDNA microarray. Although most eukaryotic mRNAs need a functional cap binding complex eIF4F for efficient 5' end-dependent scanning to initiate translation, picornaviral, hepatitis C viral, and a few cellular RNAs have been shown to be translated by internal ribosome entry, a mechanism that can operate in the presence of low levels of functional eIF4F. To identify cellular mRNAs that can be translated when eIF4F is depleted or in low abundance and that, therefore, may contain internal ribosome entry sites, mRNAs that remained associated with polysomes were isolated from human cells after infection with poliovirus and were identified by using a cDNA microarray. Approximately 200 of the 7000 mRNAs analyzed remained associated with polysomes under these conditions. Among the gene products encoded by these polysome-associated mRNAs were immediate-early transcription factors, kinases, and phosphatases of the mitogen-activated protein kinase pathways and several protooncogenes, including c-myc and Pim-1. In addition, the mRNA encoding Cyr61, a secreted factor that can promote angiogenesis and tumor growth, was selectively mobilized into polysomes when eIF4F concentrations were reduced, although its overall abundance changed only slightly. Subsequent tests confirmed the presence of internal ribosome entry sites in the 5' noncoding regions of both Cyr61 and Pim-1 mRNAs. Overall, this study suggests that diverse mRNAs whose gene products have been implicated in a variety of stress responses, including inflammation, angiogenesis, and the response to serum, can use translational initiation mechanisms that require little or no intact cap binding protein complex eIF4F. This study is described more fully in Johannes G et al.(1999) Proc Natl Acad Sci U S A 96:13118-23.", u'Set of 65 surgical specimens of human breast tumors from 42 different individuals, using complementary DNA microarrays representing 8,102 human genes. Gene expression variation patterns within this set provided a distinctive molecular portrait of each tumor. Twenty of the tumors were sampled twice, before and after a 16-week course of doxorubicin chemotherapy, and two tumors were paired with a lymph node metastasis from the same patient. Gene expression patterns in two tumor samples from the same individual were almost always more similar to each other than either was to any other sample. This study is described more fully in Perou CM et al.(2000) Nature 406:747-52', u"Diffuse large B-cell lymphoma (DLBCL), the most common subtype of non-Hodgkin's lymphoma, is clinically heterogeneous: 40% of patients respond well to current therapy and have prolonged survival, whereas the remainder succumb to the disease. We proposed that this variability in natural history reflects unrecognized molecular heterogeneity in the tumors. Using DNA microarrays, we have conducted a systematic characterization of gene expression in B-cell malignancies. Here we show that there is diversity in gene expression among the tumors of DLBCL patients, apparently reflecting the variation in tumor proliferation rate, host response and differentiation state of the tumor. We identified two molecularly distinct forms of DLBCL which had gene expression patterns indicative of different stages of B-cell differentiation. One type expressed genes characteristic of germinal center B cells ('germinal center B-like DLBCL'); the second type expressed genes normally induced during in vitro activation of peripheral blood B cells ('activated B-like DLBCL'). Patients with germinal center B-like DLBCL had a significantly better overall survival than those with activated B-like DLBCL. The molecular classification of tumors on the basis of gene expression can thus identify previously undetected and clinically significant subtypes of cancer. This study is described more fully in Alizadeh AA et al.(2000) Nature 403:503-11", u'Distinctive gene expression patterns in human mammary epithelial cells and breast cancers. cDNA microarrays and a clustering algorithm were used to identify patterns of gene expression in human mammary epithelial cells growing in culture and in primary human breast tumors. Clusters of coexpressed genes identified through manipulations of mammary epithelial cells in vitro also showed consistent patterns of variation in expression among breast tumor samples. By using immunohistochemistry with antibodies against proteins encoded by a particular gene in a cluster, the identity of the cell type within the tumor specimen that contributed the observed gene expression pattern could be determined. Clusters of genes with coherent expression patterns in cultured cells and in the breast tumors samples could be related to specific features of biological variation among the samples. Two such clusters were found to have patterns that correlated with variation in cell proliferation rates and with activation of the IFN-regulated signal transduction pathway, respectively. Clusters of genes expressed by stromal cells and lymphocytes in the breast tumors also were identified in this analysis. These results support the feasibility and usefulness of this systematic approach to studying variation in gene expression patterns in human cancers as a means to dissect and classify solid tumors. This study is described more fully in Perou CM, et al. 1999. Proc Natl Acad Sci USA 96:9212-7', u'Protein microarrays for highly parallel detection and quantitation of specific proteins and antibodies in complex solutions. This study is described more fully in Haab, Dunham, and Brown, Genome Biology 2001 2(2): research0004.1-0004.13.', u'Gene amplifications and deletions frequently contribute to tumorigenesis. Characterization of these DNA copy-number changes is important for both the basic understanding of cancer and its diagnosis. Comparative genomic hybridization (CGH) was developed to survey DNA copy-number variations across a whole genome. With CGH, differentially labelled test and reference genomic DNAs are co-hybridized to normal metaphase chromosomes, and fluorescence ratios along the length of chromosomes provide a cytogenetic representation of DNA copy-number variation. CGH, however, has a limited ( approximately 20 Mb) mapping resolution, and higher-resolution techniques, such as fluorescence in situ hybridization (FISH), are prohibitively labour-intensive on a genomic scale. Array-based CGH, in which fluorescence ratios at arrayed DNA elements provide a locus-by-locus measure of DNA copy-number variation, represents another means of achieving increased mapping resolution. Published array CGH methods have relied on large genomic clone (for example BAC) array targets and have covered only a small fraction of the human genome. cDNAs representing over 30,000 radiation-hybrid (RH)-mapped human genes provide an alternative and readily available genomic resource for mapping DNA copy-number changes. Although cDNA microarrays have been used extensively to characterize variation in human gene expression, human genomic DNA is a far more complex mixture than the mRNA representation of human cells. Therefore, analysis of DNA copy-number variation using cDNA microarrays would require a sensitivity of detection an order of magnitude greater than has been routinely reported. We describe here a cDNA microarray-based CGH method, and its application to DNA copy-number variation analysis in breast cancer cell lines and tumours. This study is described more fully in Pollack JR et al.(1999) Nat Genet 23:41-6', u'Large-scale identification of secreted and membrane-associated gene products using DNA microarrays. This study is described more fully in Diehn M et al.(2000) Nat Genet 25:58-62', u'UV-B light and Gamma-ray treated plantlet.', u'High-throughput gene expression profiling has become an important tool for investigating transcriptional activity in a variety of biological samples. To date, the vast majority of these experiments have focused on specific biological processes and perturbations. Here, we have generated and analyzed gene expression from a set of samples spanning a broad range of biological conditions. Specifically, we profiled gene expression from 91 human and mouse samples across a diverse array of tissues, organs, and cell lines. Because these samples predominantly come from the normal physiological state in the human and mouse, this dataset represents a preliminary, but substantial, description of the normal mammalian transcriptome. We have used this dataset to illustrate methods of mining these data, and to reveal insights into molecular and physiological gene function, mechanisms of transcriptional regulation, disease etiology, and comparative genomics. Finally, to allow the scientific community to use this resource, we have built a free and publicly accessible website (<a href="http://expression.gnf.org" target="_blank">http://expression.gnf.org</a>) that integrates data visualization and curation of current gene annotations. The terms of the Novartis data release policy prevent us from redistributing .cel files for this experiment, therefore there is no raw data, we hope to be able to release this in future.', u'SAGE libraries from cultured, differentiated keratinocytes and human epidermis, both normal and affected by actinic keratosis Keywords = Keratinocyte, Epidermis, Homo sapiens, Actinic Keratosis, TNF alpha', u'A Saccharomyces cerevisiae population was cultured for many generations under conditions to which it is not optimally adapted. These experiments were designed to investigate adaptive evolution under natural selection. This study is described in more detail in Ferea TL, et al. 1999. Proc Natl Acad Sci USA 96:9721-6', u'Yeast cells were blocked in telophase using a cdc15-2 temperature senstive mutant at restrictive temperature. The culture was then shifted to permissive temperature (25oC), and released into the cell cycle. Samples were then taken during the course of almost three full cell cycles. This study is described in more detail in Spellman PT et al.(1998) Mol Biol Cell 9:3273-97', u'The arrays were used to study the effects of modifying phosphate levels in yeast, either through mutation, or chemical manipulation. This study is described in more detail in Ogawa N et al.(2000) Mol Biol Cell 11:4309-21', u'Human retinal and RPE SAGE libraries. Profile of the genes expressed in the human peripheral retina, macula, and retinal pigment epithelium determined through serial analysis of gene expression (SAGE). Keywords: other', u'Analysis of topoisomerase function in bacterial replication fork movement: use of DNA microarrays. We used DNA microarrays of the Escherichia coli genome to trace the progression of chromosomal replication forks in synchronized cells. We found that both DNA gyrase and topoisomerase IV (topo IV) promote replication fork progression. When both enzymes were inhibited, the replication fork stopped rapidly. The elongation rate with topo IV alone was 1/3 of normal. Genetic data confirmed and extended these results. Inactivation of gyrase alone caused a slow stop of replication. Topo IV activity was sufficient to prevent accumulation of (+) supercoils in plasmid DNA in vivo, suggesting that topo IV can promote replication by removing (+) supercoils in front of the chromosomal fork. This study is detailed in Khodursky AB et al.(2000) Proc Natl Acad Sci U S A 97:9419-24 Keywords: other', u"Helicobacter pylori colonizes the stomach of half of the world's population, causing a wide spectrum of disease ranging from asymptomatic gastritis to ulcers to gastric cancer. Although the basis for these diverse clinical outcomes is not understood, more severe disease is associated with strains harboring a pathogenicity island. To characterize the genetic diversity of more and less virulent strains, we examined the genomic content of 15 H. pylori clinical isolates by using a whole genome H. pylori DNA microarray. We found that a full 22% of H. pylori genes are dispensable in one or more strains, thus defining a minimal functional core of 1281 H. pylori genes. While the core genes encode most metabolic and cellular processes, the strain-specific genes include genes unique to H. pylori, restriction modification genes, transposases, and genes encoding cell surface proteins, which may aid the bacteria under specific circumstances during their long-term infection of genetically diverse hosts. We observed distinct patterns of the strain-specific gene distribution along the chromosome, which may result from different mechanisms of gene acquisition and loss. Among the strain-specific genes, we have found a class of candidate virulence genes identified by their coinheritance with the pathogenicity island."])
		self.checkEqual([ob.name for thing in [c.organism_ for c in q.all()] for ob in thing], [u'Plasmodium falciparum 3D7', u'Bos taurus', u'Medicago truncatula', u'Schizosaccharomyces pombe', u'Schizosaccharomyces pombe', u'Schizosaccharomyces pombe', u'Schizosaccharomyces pombe', u'Schizosaccharomyces pombe', u'Schizosaccharomyces pombe', u'Mus musculus', u'Homo sapiens', u'Homo sapiens', u'Homo sapiens', u'Homo sapiens', u'Homo sapiens', u'Homo sapiens', u'Homo sapiens', u'Arabidopsis thaliana', u'Homo sapiens', u'Homo sapiens', u'Saccharomyces cerevisiae', u'Saccharomyces cerevisiae', u'Saccharomyces cerevisiae', u'Homo sapiens', u'Escherichia coli', u'Helicobacter pylori'])
		self.checkEqual([ob.name for thing in [c.experiment_types for c in q.all()] for ob in thing], [u'other', u'transcription profiling by SAGE', u'unknown experiment type', u'transcription profiling by array', u'transcription profiling by array', u'transcription profiling by array', u'transcription profiling by array', u'transcription profiling by array', u'transcription profiling by array', u'transcription profiling by SAGE', u'unknown experiment type', u'unknown experiment type', u'unknown experiment type', u'unknown experiment type', u'unknown experiment type', u'unknown experiment type', u'unknown experiment type', u'unknown experiment type', u'transcription profiling by array', u'transcription profiling by SAGE', u'unknown experiment type', u'unknown experiment type', u'unknown experiment type', u'transcription profiling by SAGE', u'other', u'unknown experiment type'])
		self.checkEqual([ob.name for thing in [c.experiment_designs for c in q.all()] for ob in thing], [u'time series', u'time series', u'time series', u'genetic modification', u'time series', u'time series', u'disease state', u'organism part comparison'])
		print ([ob.name for thing in [c.data_formats for c in q.all()] for ob in thing])
		#self.checkEqual([ob.name for thing in [c.data_formats for c in q.all()] for ob in thing], [u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData', u'scan', u'processedDataMatrix', u'scan', u'processedDataMatrix', u'scan', u'processedDataMatrix', u'scan', u'processedDataMatrix', u'scan', u'processedDataMatrix', u'scan', u'processedDataMatrix', u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData', u'normalization', u'processedData'])
		self.checkEqual([c.submission_date for c in q.all()], [u'2002-11-08', u'2002-12-10', u'2002-07-23', None, None, None, None, None, None, u'2002-05-22', u'2002-06-23', u'2002-06-23', u'2002-06-23', u'2002-06-17', u'2002-05-22', u'2002-05-22', u'2002-05-22', u'2002-06-17', None, u'2002-02-14', u'2002-02-08', u'2002-02-08', u'2002-02-08', u'2001-10-02', u'2001-07-29', u'2001-07-24'])
		self.checkEqual([c.release_date for c in q.all()], [u'2002-12-22', u'2002-12-09', u'2002-11-19', u'2002-08-05', u'2002-08-05', u'2002-08-05', u'2002-08-05', u'2002-08-05', u'2002-08-05', u'2002-07-18', u'2002-06-27', u'2002-06-27', u'2002-06-27', u'2002-06-27', u'2002-06-27', u'2002-06-27', u'2002-06-27', u'2002-06-19', u'2002-03-19', u'2002-02-21', u'2002-02-11', u'2002-02-11', u'2002-02-11', u'2001-10-03', u'2001-07-30', u'2001-07-25'])

		q = session.query(array_express.DataFormat)
		self.checkEqual([c.name for c in q.all()], [u'normalization', u'processedData', u'scan', u'processedDataMatrix'])

		q = session.query(array_express.ExperimentDesign)
		self.checkEqual([c.name for c in q.all()] , [u'time series', u'genetic modification', u'disease state', u'organism part comparison'])
		
		q = session.query(array_express.ExperimentType)
		self.checkEqual([c.name for c in q.all()], [u'other', u'transcription profiling by SAGE', u'unknown experiment type', u'transcription profiling by array'])

		q = session.query(array_express.Organism)
		self.checkEqual([c.name for c in q.all()], [u'Plasmodium falciparum 3D7', u'Bos taurus', u'Medicago truncatula', u'Schizosaccharomyces pombe', u'Mus musculus', u'Homo sapiens', u'Arabidopsis thaliana', u'Saccharomyces cerevisiae', u'Escherichia coli', u'Helicobacter pylori'])

		q = session.query(array_express.Sample)
		self.assertEqual(sorted([c.name for c in q.all()])[:10], 
			[u'GSM1026 1', u'GSM1026 2', u'GSM1028 1', u'GSM1028 2', u'GSM1031 1', u'GSM1031 2', u'GSM1036 1', u'GSM1036 2', u'GSM1056 1', u'GSM1056 2'])
		self.checkEqual(sorted([c.assay for c in q.all()])[:10], 
			[u'GSM1026', u'GSM1026', u'GSM1028', u'GSM1028', u'GSM1031', u'GSM1031', u'GSM1036', u'GSM1036', u'GSM1056', u'GSM1056'])
		self.checkEqual([ob.name for thing in [c.characteristics for c in q.all()] for ob in thing], [u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Channel', u'Channel', u'Genotype', u'Organism', u'Channel', u'Channel', u'Genotype', u'Organism', u'Channel', u'Channel', u'Genotype', u'Organism', u'Channel', u'Channel', u'Genotype', u'Organism', u'Channel', u'Channel', u'Genotype', u'Organism', u'Channel', u'Channel', u'Genotype', u'Organism', u'Channel', u'Channel', u'Genotype', u'Organism', u'Channel', u'Channel', u'Genotype', u'Organism', u'Channel', u'Channel', u'Genotype', u'Organism', u'Channel', u'Channel', u'Genotype', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Genotype', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'OrganismPart', u'Organism', u'DevelopmentalStage', u'Sex', u'Age', u'TimeUnit', u'InitialTimePoint', u'DiseaseState', u'CellLine', u'IndividualGeneticCharacteristics', u'CellType', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Organism'])
		self.checkEqual([ob.name for thing in [c.variables for c in q.all()] for ob in thing], [u'genotype', u'genotype', u'genotype', u'genotype', u'genotype', u'genotype', u'genotype', u'genotype', u'genotype', u'genotype', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'genotype', u'genotype', u'genotype', u'genotype', u'genotype', u'genotype', u'genotype', u'genotype', u'genotype', u'genotype', u'genotype', u'genotype'])
		
		q = session.query(array_express.Extract)
		self.assertEqual(len([c.sample_id for c in q.all()]), 1373)
		self.checkEqual([c.name for c in q.all()], [u'GSM1725 extract 1', u'GSM1725 extract 2', u'GSM1722 extract 2', u'GSM1722 extract 1', u'GSM1717 extract 1', u'GSM1717 extract 2', u'GSM1723 extract 2', u'GSM1723 extract 1', u'GSM1716 extract 1', u'GSM1716 extract 2', u'GSM1718 extract 2', u'GSM1718 extract 1', u'GSM1719 extract 1', u'GSM1719 extract 2', u'GSM1720 extract 2', u'GSM1720 extract 1', u'GSM1721 extract 2', u'GSM1721 extract 1', u'GSM1724 extract 1', u'GSM1724 extract 2', u'GSM3037 extract 1', u'GSM3036 extract 1', u'GSM1714 extract 2', u'GSM1714 extract 1', u'GSM1713 extract 1', u'GSM1713 extract 2', u'GSM1712 extract 2', u'GSM1712 extract 1', u'GSM1711 extract 2', u'GSM1711 extract 1', u'GSM1710 extract 2', u'GSM1710 extract 1', u'GSM1709 extract 1', u'GSM1709 extract 2', u'GSM1708 extract 1', u'GSM1708 extract 2', u'GSM1707 extract 1', u'GSM1707 extract 2', u'GSM1706 extract 1', u'GSM1706 extract 2', u'GSM1705 extract 2', u'GSM1705 extract 1', u'GSM1704 extract 2', u'GSM1704 extract 1', u'GSM1703 extract 1', u'GSM1703 extract 2', u'GSM1702 extract 1', u'GSM1702 extract 2', u'GSM1701 extract 1', u'GSM1701 extract 2', u'GSM1700 extract 2', u'GSM1700 extract 1', u'GSM1699 extract 2', u'GSM1699 extract 1', u'GSM1698 extract 2', u'GSM1698 extract 1', u'GSM1697 extract 1', u'GSM1697 extract 2', u'GSM1696 extract 1', u'GSM1696 extract 2', u'GSM1695 extract 2', u'GSM1695 extract 1', u'GSM1694 extract 1', u'GSM1694 extract 2', u'GSM1693 extract 1', u'GSM1693 extract 2', u'GSM1692 extract 1', u'GSM1692 extract 2', u'GSM1691 extract 2', u'GSM1691 extract 1', u'GSM1690 extract 2', u'GSM1690 extract 1', u'GSM1689 extract 2', u'GSM1689 extract 1', u'GSM1688 extract 1', u'GSM1688 extract 2', u'GSM1687 extract 1', u'GSM1687 extract 2', u'GSM1686 extract 1', u'GSM1686 extract 2', u'GSM1685 extract 1', u'GSM1685 extract 2', u'GSM505 extract 2', u'GSM505 extract 1', u'GSM504 extract 1', u'GSM504 extract 2', u'GSM503 extract 2', u'GSM503 extract 1', u'GSM502 extract 1', u'GSM502 extract 2', u'GSM501 extract 1', u'GSM501 extract 2', u'GSM500 extract 1', u'GSM500 extract 2', u'GSM499 extract 1', u'GSM499 extract 2', u'GSM498 extract 1', u'GSM498 extract 2', u'GSM497 extract 2', u'GSM497 extract 1', u'GSM496 extract 2', u'GSM496 extract 1', u'GSM495 extract 1', u'GSM495 extract 2', u'GSM494 extract 1', u'GSM494 extract 2', u'GSM493 extract 2', u'GSM493 extract 1', u'GSM492 extract 2', u'GSM492 extract 1', u'GSM491 extract 2', u'GSM491 extract 1', u'GSM490 extract 2', u'GSM490 extract 1', u'GSM489 extract 2', u'GSM489 extract 1', u'GSM488 extract 1', u'GSM488 extract 2', u'GSM487 extract 1', u'GSM487 extract 2', u'GSM486 extract 2', u'GSM486 extract 1', u'GSM485 extract 1', u'GSM485 extract 2', u'GSM484 extract 1', u'GSM484 extract 2', u'GSM483 extract 2', u'GSM483 extract 1', u'GSM482 extract 2', u'GSM482 extract 1', u'GSM481 extract 2', u'GSM481 extract 1', u'GSM480 extract 2', u'GSM480 extract 1', u'GSM479 extract 2', u'GSM479 extract 1', u'GSM478 extract 2', u'GSM478 extract 1', u'GSM477 extract 1', u'GSM477 extract 2', u'GSM476 extract 1', u'GSM476 extract 2', u'GSM475 extract 2', u'GSM475 extract 1', u'GSM474 extract 2', u'GSM474 extract 1', u'GSM473 extract 1', u'GSM473 extract 2', u'GSM472 extract 2', u'GSM472 extract 1', u'GSM471 extract 1', u'GSM471 extract 2', u'GSM470 extract 1', u'GSM470 extract 2', u'GSM1927 extract 2', u'GSM1927 extract 1', u'GSM1926 extract 1', u'GSM1926 extract 2', u'GSM1925 extract 2', u'GSM1925 extract 1', u'GSM1924 extract 1', u'GSM1924 extract 2', u'GSM1923 extract 1', u'GSM1923 extract 2', u'GSM1922 extract 2', u'GSM1922 extract 1', u'GSM1921 extract 1', u'GSM1921 extract 2', u'GSM1920 extract 2', u'GSM1920 extract 1', u'GSM1919 extract 2', u'GSM1919 extract 1', u'GSM1918 extract 2', u'GSM1918 extract 1', u'GSM1917 extract 1', u'GSM1917 extract 2', u'GSM1916 extract 1', u'GSM1916 extract 2', u'GSM1915 extract 1', u'GSM1915 extract 2', u'GSM1914 extract 1', u'GSM1914 extract 2', u'GSM1913 extract 2', u'GSM1913 extract 1', u'GSM1912 extract 1', u'GSM1912 extract 2', u'GSM1911 extract 2', u'GSM1911 extract 1', u'GSM1910 extract 1', u'GSM1910 extract 2', u'GSM1909 extract 1', u'GSM1909 extract 2', u'GSM1908 extract 1', u'GSM1908 extract 2', u'GSM1907 extract 2', u'GSM1907 extract 1', u'GSM1906 extract 1', u'GSM1906 extract 2', u'GSM1905 extract 2', u'GSM1905 extract 1', u'GSM1904 extract 1', u'GSM1904 extract 2', u'GSM1903 extract 2', u'GSM1903 extract 1', u'GSM1902 extract 1', u'GSM1902 extract 2', u'GSM1901 extract 1', u'GSM1901 extract 2', u'GSM1900 extract 1', u'GSM1900 extract 2', u'GSM1899 extract 1', u'GSM1899 extract 2', u'GSM1898 extract 1', u'GSM1898 extract 2', u'GSM1897 extract 1', u'GSM1897 extract 2', u'GSM1896 extract 2', u'GSM1896 extract 1', u'GSM1895 extract 2', u'GSM1895 extract 1', u'GSM1894 extract 2', u'GSM1894 extract 1', u'GSM1893 extract 1', u'GSM1893 extract 2', u'GSM1892 extract 2', u'GSM1892 extract 1', u'GSM1891 extract 1', u'GSM1891 extract 2', u'GSM1890 extract 1', u'GSM1890 extract 2', u'GSM1889 extract 2', u'GSM1889 extract 1', u'GSM1888 extract 2', u'GSM1888 extract 1', u'GSM1887 extract 1', u'GSM1887 extract 2', u'GSM1886 extract 1', u'GSM1886 extract 2', u'GSM1885 extract 1', u'GSM1885 extract 2', u'GSM1884 extract 1', u'GSM1884 extract 2', u'GSM1883 extract 1', u'GSM1883 extract 2', u'GSM1882 extract 1', u'GSM1882 extract 2', u'GSM1881 extract 1', u'GSM1881 extract 2', u'GSM1880 extract 1', u'GSM1880 extract 2', u'GSM1879 extract 1', u'GSM1879 extract 2', u'GSM1878 extract 1', u'GSM1878 extract 2', u'GSM1877 extract 1', u'GSM1877 extract 2', u'GSM1876 extract 1', u'GSM1876 extract 2', u'GSM1875 extract 1', u'GSM1875 extract 2', u'GSM1874 extract 2', u'GSM1874 extract 1', u'GSM1873 extract 1', u'GSM1873 extract 2', u'GSM1872 extract 2', u'GSM1872 extract 1', u'GSM1871 extract 2', u'GSM1871 extract 1', u'GSM1870 extract 1', u'GSM1870 extract 2', u'GSM1869 extract 2', u'GSM1869 extract 1', u'GSM1868 extract 2', u'GSM1868 extract 1', u'GSM1867 extract 2', u'GSM1867 extract 1', u'GSM1866 extract 1', u'GSM1866 extract 2', u'GSM1865 extract 1', u'GSM1865 extract 2', u'GSM1864 extract 1', u'GSM1864 extract 2', u'GSM1863 extract 1', u'GSM1863 extract 2', u'GSM1862 extract 2', u'GSM1862 extract 1', u'GSM1861 extract 2', u'GSM1861 extract 1', u'GSM1860 extract 2', u'GSM1860 extract 1', u'GSM1859 extract 1', u'GSM1859 extract 2', u'GSM1858 extract 2', u'GSM1858 extract 1', u'GSM1857 extract 1', u'GSM1857 extract 2', u'GSM1856 extract 2', u'GSM1856 extract 1', u'GSM1855 extract 1', u'GSM1855 extract 2', u'GSM1854 extract 2', u'GSM1854 extract 1', u'GSM1853 extract 1', u'GSM1853 extract 2', u'GSM1852 extract 2', u'GSM1852 extract 1', u'GSM1851 extract 1', u'GSM1851 extract 2', u'GSM1850 extract 2', u'GSM1850 extract 1', u'GSM1849 extract 2', u'GSM1849 extract 1', u'GSM1848 extract 1', u'GSM1848 extract 2', u'GSM1847 extract 1', u'GSM1847 extract 2', u'GSM1846 extract 2', u'GSM1846 extract 1', u'GSM1845 extract 2', u'GSM1845 extract 1', u'GSM1844 extract 1', u'GSM1844 extract 2', u'GSM574 extract 1', u'GSM573 extract 1', u'GSM572 extract 1', u'GSM571 extract 1', u'GSM569 extract 2', u'GSM569 extract 1', u'GSM568 extract 2', u'GSM568 extract 1', u'GSM567 extract 2', u'GSM567 extract 1', u'GSM566 extract 1', u'GSM566 extract 2', u'GSM554 extract 2', u'GSM554 extract 1', u'GSM553 extract 2', u'GSM553 extract 1', u'GSM552 extract 2', u'GSM552 extract 1', u'GSM551 extract 1', u'GSM551 extract 2', u'GSM550 extract 1', u'GSM550 extract 2', u'GSM530 extract 2', u'GSM530 extract 1', u'GSM529 extract 2', u'GSM529 extract 1', u'GSM528 extract 2', u'GSM528 extract 1', u'GSM527 extract 1', u'GSM527 extract 2', u'GSM526 extract 1', u'GSM526 extract 2', u'GSM525 extract 1', u'GSM525 extract 2', u'GSM524 extract 2', u'GSM524 extract 1', u'GSM523 extract 1', u'GSM523 extract 2', u'GSM522 extract 2', u'GSM522 extract 1', u'GSM521 extract 1', u'GSM521 extract 2', u'GSM520 extract 2', u'GSM520 extract 1', u'GSM518 extract 1', u'GSM518 extract 2', u'GSM901 extract 1', u'GSM901 extract 2', u'GSM1079 extract 2', u'GSM1079 extract 1', u'GSM1072 extract 2', u'GSM1072 extract 1', u'GSM1099 extract 2', u'GSM1099 extract 1', u'GSM1098 extract 2', u'GSM1098 extract 1', u'GSM1109 extract 1', u'GSM1109 extract 2', u'GSM909 extract 2', u'GSM909 extract 1', u'GSM1097 extract 2', u'GSM1097 extract 1', u'GSM1062 extract 2', u'GSM1062 extract 1', u'GSM1096 extract 2', u'GSM1096 extract 1', u'GSM897 extract 2', u'GSM897 extract 1', u'GSM1056 extract 1', u'GSM1056 extract 2', u'GSM1095 extract 1', u'GSM1095 extract 2', u'GSM898 extract 2', u'GSM898 extract 1', u'GSM1092 extract 1', u'GSM1092 extract 2', u'GSM900 extract 1', u'GSM900 extract 2', u'GSM1091 extract 2', u'GSM1091 extract 1', u'GSM911 extract 2', u'GSM911 extract 1', u'GSM1086 extract 2', u'GSM1086 extract 1', u'GSM1036 extract 1', u'GSM1036 extract 2', u'GSM1108 extract 1', u'GSM1108 extract 2', u'GSM1031 extract 2', u'GSM1031 extract 1', u'GSM1028 extract 1', u'GSM1028 extract 2', u'GSM896 extract 1', u'GSM896 extract 2', u'GSM1026 extract 1', u'GSM1026 extract 2', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'8h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'8h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'2h replicate 1', u'7h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'1h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'3h replicate 1', u'0h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'6h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'2h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'7h replicate 1', u'1h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'-14ha replicate 1', u'-14hb replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'4h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'12h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'-14ha replicate 1', u'5h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'4h replicate 1', u'0h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'10h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'3h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'10h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'6h replicate 1', u'12h replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'pool replicate 1', u'5h replicate 1', u'GSM2086 extract 1', u'GSM2086 extract 2', u'GSM2085 extract 2', u'GSM2085 extract 1', u'GSM2084 extract 2', u'GSM2084 extract 1', u'GSM2083 extract 2', u'GSM2083 extract 1', u'GSM2082 extract 2', u'GSM2082 extract 1', u'GSM2081 extract 1', u'GSM2081 extract 2', u'GSM2080 extract 1', u'GSM2080 extract 2', u'GSM2079 extract 2', u'GSM2079 extract 1', u'GSM2078 extract 2', u'GSM2078 extract 1', u'GSM2077 extract 1', u'GSM2077 extract 2', u'GSM2076 extract 1', u'GSM2076 extract 2', u'GSM2075 extract 1', u'GSM2075 extract 2', u'GSM2074 extract 1', u'GSM2074 extract 2', u'GSM2073 extract 2', u'GSM2073 extract 1', u'GSM2072 extract 1', u'GSM2072 extract 2', u'GSM2071 extract 1', u'GSM2071 extract 2', u'GSM2070 extract 2', u'GSM2070 extract 1', u'GSM2069 extract 2', u'GSM2069 extract 1', u'GSM2068 extract 2', u'GSM2068 extract 1', u'GSM2067 extract 1', u'GSM2067 extract 2', u'GSM2066 extract 1', u'GSM2066 extract 2', u'GSM2065 extract 1', u'GSM2065 extract 2', u'GSM2064 extract 2', u'GSM2064 extract 1', u'GSM2063 extract 2', u'GSM2063 extract 1', u'GSM2062 extract 2', u'GSM2062 extract 1', u'GSM2061 extract 1', u'GSM2061 extract 2', u'GSM2060 extract 2', u'GSM2060 extract 1', u'GSM2059 extract 1', u'GSM2059 extract 2', u'GSM2058 extract 2', u'GSM2058 extract 1', u'GSM2057 extract 1', u'GSM2057 extract 2', u'GSM2056 extract 2', u'GSM2056 extract 1', u'GSM2055 extract 2', u'GSM2055 extract 1', u'GSM2054 extract 1', u'GSM2054 extract 2', u'GSM2053 extract 1', u'GSM2053 extract 2', u'GSM2052 extract 1', u'GSM2052 extract 2', u'GSM2051 extract 1', u'GSM2051 extract 2', u'GSM2050 extract 2', u'GSM2050 extract 1', u'GSM2049 extract 1', u'GSM2049 extract 2', u'GSM2048 extract 1', u'GSM2048 extract 2', u'GSM2047 extract 1', u'GSM2047 extract 2', u'GSM2046 extract 2', u'GSM2046 extract 1', u'GSM2045 extract 2', u'GSM2045 extract 1', u'GSM2044 extract 1', u'GSM2044 extract 2', u'GSM2043 extract 2', u'GSM2043 extract 1', u'GSM2042 extract 1', u'GSM2042 extract 2', u'GSM2041 extract 2', u'GSM2041 extract 1', u'GSM2040 extract 2', u'GSM2040 extract 1', u'GSM2039 extract 2', u'GSM2039 extract 1', u'GSM2038 extract 2', u'GSM2038 extract 1', u'GSM2037 extract 2', u'GSM2037 extract 1', u'GSM2036 extract 1', u'GSM2036 extract 2', u'GSM2035 extract 1', u'GSM2035 extract 2', u'GSM2034 extract 2', u'GSM2034 extract 1', u'GSM2033 extract 1', u'GSM2033 extract 2', u'GSM2032 extract 1', u'GSM2032 extract 2', u'GSM2031 extract 1', u'GSM2031 extract 2', u'GSM2030 extract 2', u'GSM2030 extract 1', u'GSM2029 extract 1', u'GSM2029 extract 2', u'GSM2028 extract 1', u'GSM2028 extract 2', u'GSM2027 extract 1', u'GSM2027 extract 2', u'GSM2026 extract 2', u'GSM2026 extract 1', u'GSM2025 extract 1', u'GSM2025 extract 2', u'GSM2024 extract 2', u'GSM2024 extract 1', u'GSM2023 extract 2', u'GSM2023 extract 1', u'GSM2022 extract 1', u'GSM2022 extract 2', u'GSM2021 extract 2', u'GSM2021 extract 1', u'GSM2020 extract 1', u'GSM2020 extract 2', u'GSM2019 extract 2', u'GSM2019 extract 1', u'GSM2018 extract 1', u'GSM2018 extract 2', u'GSM2017 extract 1', u'GSM2017 extract 2', u'GSM2016 extract 1', u'GSM2016 extract 2', u'GSM2015 extract 1', u'GSM2015 extract 2', u'GSM2014 extract 2', u'GSM2014 extract 1', u'GSM2013 extract 1', u'GSM2013 extract 2', u'GSM2012 extract 2', u'GSM2012 extract 1', u'GSM2011 extract 2', u'GSM2011 extract 1', u'GSM2010 extract 2', u'GSM2010 extract 1', u'GSM2009 extract 2', u'GSM2009 extract 1', u'GSM2008 extract 2', u'GSM2008 extract 1', u'GSM2007 extract 1', u'GSM2007 extract 2', u'GSM2006 extract 2', u'GSM2006 extract 1', u'GSM2005 extract 1', u'GSM2005 extract 2', u'GSM2004 extract 2', u'GSM2004 extract 1', u'GSM2003 extract 1', u'GSM2003 extract 2', u'GSM2002 extract 2', u'GSM2002 extract 1', u'GSM2001 extract 1', u'GSM2001 extract 2', u'GSM2000 extract 2', u'GSM2000 extract 1', u'GSM1999 extract 1', u'GSM1999 extract 2', u'GSM1998 extract 2', u'GSM1998 extract 1', u'GSM1997 extract 1', u'GSM1997 extract 2', u'GSM1996 extract 1', u'GSM1996 extract 2', u'GSM1995 extract 2', u'GSM1995 extract 1', u'GSM1994 extract 2', u'GSM1994 extract 1', u'GSM1993 extract 1', u'GSM1993 extract 2', u'GSM1992 extract 2', u'GSM1992 extract 1', u'GSM1991 extract 2', u'GSM1991 extract 1', u'GSM1990 extract 1', u'GSM1990 extract 2', u'GSM1989 extract 1', u'GSM1989 extract 2', u'GSM1988 extract 1', u'GSM1988 extract 2', u'GSM1987 extract 1', u'GSM1987 extract 2', u'GSM1986 extract 2', u'GSM1986 extract 1', u'GSM1985 extract 2', u'GSM1985 extract 1', u'GSM1984 extract 1', u'GSM1984 extract 2', u'GSM1983 extract 2', u'GSM1983 extract 1', u'GSM1982 extract 2', u'GSM1982 extract 1', u'GSM1981 extract 1', u'GSM1981 extract 2', u'GSM1980 extract 1', u'GSM1980 extract 2', u'GSM1979 extract 1', u'GSM1979 extract 2', u'GSM1978 extract 2', u'GSM1978 extract 1', u'GSM1977 extract 1', u'GSM1977 extract 2', u'GSM1976 extract 2', u'GSM1976 extract 1', u'GSM1975 extract 1', u'GSM1975 extract 2', u'GSM1974 extract 1', u'GSM1974 extract 2', u'GSM1973 extract 2', u'GSM1973 extract 1', u'GSM1972 extract 2', u'GSM1972 extract 1', u'GSM1971 extract 1', u'GSM1971 extract 2', u'GSM1970 extract 1', u'GSM1970 extract 2', u'GSM1969 extract 1', u'GSM1969 extract 2', u'GSM1968 extract 1', u'GSM1968 extract 2', u'GSM1967 extract 1', u'GSM1967 extract 2', u'GSM1966 extract 1', u'GSM1966 extract 2', u'GSM1965 extract 1', u'GSM1965 extract 2', u'GSM1964 extract 1', u'GSM1964 extract 2', u'GSM1963 extract 1', u'GSM1963 extract 2', u'GSM1962 extract 2', u'GSM1962 extract 1', u'GSM1961 extract 1', u'GSM1961 extract 2', u'GSM1960 extract 1', u'GSM1960 extract 2', u'GSM1959 extract 2', u'GSM1959 extract 1', u'GSM1958 extract 2', u'GSM1958 extract 1', u'GSM1957 extract 2', u'GSM1957 extract 1', u'GSM1956 extract 1', u'GSM1956 extract 2', u'GSM1955 extract 2', u'GSM1955 extract 1', u'GSM1954 extract 1', u'GSM1954 extract 2', u'GSM1684 extract 1', u'GSM1683 extract 1', u'GSM1682 extract 1', u'GSM2582 extract 1', u'GSM2581 extract 1', u'GSM2580 extract 1', u'GSM2579 extract 1', u'1h, added nitrogen', u'pool', u'pool', u'pool', u'pool', u'-14h, added nitrogen', u'pool', u'pool', u'pool', u'pool', u'0h, added nitrogen', u'pool', u'pool', u'pool', u'pool', u'-14h, added nitrogen', u'pool', u'pool', u'pool', u'pool', u'GSM2177 extract 1', u'GSM2177 extract 2', u'GSM2176 extract 1', u'GSM2176 extract 2', u'GSM2175 extract 2', u'GSM2175 extract 1', u'GSM2145 extract 1', u'GSM2145 extract 2', u'GSM2144 extract 1', u'GSM2144 extract 2', u'GSM1728 extract 2', u'GSM1728 extract 1', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt 7h', u'wt 12h', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt 4h', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt 0ha', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt 2h', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt 5h', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt 10hb', u'wt 8h', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt 3h', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt 12h__S:wt_meiosis_12h', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt 9h', u'wt 12h__S:wt_meiosis_10h', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt 1h', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt pool', u'wt 6h', u'GSM1715 extract 2', u'GSM1715 extract 1', u'GSM770 extract 1', u'GSM769 extract 1', u'GSM768 extract 1', u'caudate_extract_2', u'GSM1950 extract 2', u'GSM1950 extract 1', u'GSM1953 extract 1', u'GSM1953 extract 2', u'GSM1928 extract 2', u'GSM1928 extract 1', u'GSM1945 extract 2', u'GSM1945 extract 1', u'GSM1952 extract 2', u'GSM1952 extract 1', u'GSM1929 extract 1', u'GSM1929 extract 2', u'GSM1934 extract 2', u'GSM1934 extract 1', u'GSM1948 extract 2', u'GSM1948 extract 1', u'GSM1951 extract 1', u'GSM1951 extract 2', u'GSM1932 extract 2', u'GSM1932 extract 1', u'GSM1935 extract 1', u'GSM1935 extract 2', u'GSM1936 extract 1', u'GSM1936 extract 2', u'GSM1949 extract 1', u'GSM1949 extract 2', u'GSM1944 extract 2', u'GSM1944 extract 1', u'GSM1938 extract 1', u'GSM1938 extract 2', u'GSM1930 extract 1', u'GSM1930 extract 2', u'GSM1939 extract 2', u'GSM1939 extract 1', u'GSM1943 extract 1', u'GSM1943 extract 2', u'GSM1947 extract 1', u'GSM1947 extract 2', u'GSM1942 extract 1', u'GSM1942 extract 2', u'GSM1940 extract 2', u'GSM1940 extract 1', u'GSM1937 extract 2', u'GSM1937 extract 1', u'GSM1941 extract 2', u'GSM1941 extract 1', u'GSM1931 extract 2', u'GSM1931 extract 1', u'GSM1946 extract 1', u'GSM1946 extract 2', u'GSM1933 extract 2', u'GSM1933 extract 1', u'GSM905 extract 2', u'GSM905 extract 1', u'GSM904 extract 1', u'GSM904 extract 2', u'GSM1094 extract 2', u'GSM1094 extract 1', u'GSM1093 extract 2', u'GSM1093 extract 1', u'GSM866 extract 2', u'GSM866 extract 1', u'GSM894 extract 2', u'GSM894 extract 1', u'GSM869 extract 1', u'GSM869 extract 2', u'GSM886 extract 1', u'GSM886 extract 2', u'GSM867 extract 1', u'GSM867 extract 2', u'GSM870 extract 2', u'GSM870 extract 1', u'GSM871 extract 1', u'GSM871 extract 2', u'GSM885 extract 1', u'GSM885 extract 2', u'GSM2087 extract 1', u'GSM1741 extract 1', u'GSM1740 extract 1', u'GSM1125 extract 1', u'GSM1124 extract 1', u'GSM1123 extract 1', u'GSM1122 extract 1', u'GSM1121 extract 1'])

		q = session.query(array_express.Variable)
		self.checkEqual([c.name for c in q.all()], [u'genotype', u'genotype', u'genotype', u'genotype', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'sampling interval', u'genotype'])
		self.checkEqual([c.value for c in q.all()], [None, u'h90_atf31::kanR', u'h90', u'h90_atf21::kanR', None, u'8', u'2', u'7', u'-14', u'6', u'1', u'4', u'12', u'5', u'10', u'3', u'0', u'11', u'h+/h+ ade6-M210/ade6-M216'])
		print([c.unit for c in q.all()])

		q = session.query(array_express.Characteristic)
		#print [c.name for c in q.all()]
		self.checkEqual([c.name for c in q.all()],[u'Organism', u'Organism', u'Organism', u'Channel', u'Channel', u'Genotype', u'Organism', u'Channel', u'Genotype', u'Genotype', u'Organism', u'Organism', u'Genotype', u'Organism', u'Organism', u'Genotype', u'Genotype', u'Organism', u'OrganismPart', u'DevelopmentalStage', u'Sex', u'Age', u'TimeUnit', u'InitialTimePoint', u'DiseaseState', u'CellLine', u'IndividualGeneticCharacteristics', u'CellType', u'Organism'])
		self.checkEqual([c.value for c in q.all()],[u'Homo sapiens', u'Bos taurus', u'Helicobacter pylori', u'Cy3', None, u'h90_atf31::kanR', u'Schizosaccharomyces pombe', u'Cy5', u'h90', u'h90_atf21::kanR', u'Escherichia coli', u'Saccharomyces cerevisiae', u'h+/h+ pat1-114/pat1-114 ade6-M210/ade6-M216', u'Mus musculus', u'Plasmodium falciparum 3D7', u'h+/h+ ade6-M210/ade6-M216', u'h+/h- ade6-M210/ade6-M216', u'Arabidopsis thaliana', u'caudate nucleus', u'adult', u'mixed_sex', u'84', u'seconds', u'birth', u'cancer', u'Daudi', u'cancer', u'hepatocyte', u'Medicago truncatula'])

	def checkEqual(self, L1, L2):
		self.assertEqual(sorted(L1), sorted(L2))
